Fix out-of-bounds tensor indexing by clamping 't_index' to valid range

Jaewoo-Kim-Rice opened the pull request:
# Summary

Fix occasional out-of-bounds indexing in LocalSliceLgCollapseP caused by floating-point rounding differences across GPUs.

# Root Cause

t_index was computed using continuous time values and torch.floor, which could produce indices slightly larger than l_dt (e.g., 127360 for a tensor of size 127360) due to floating-point precision differences.

# Fix

Clamp all computed indices to the valid range before use:

t_index = torch.clamp(t_index, 0, l_dt - 1)

# Result

Prevents IndexError / CUDA device-side assert while preserving identical behavior for valid indices and ensuring stable, hardware-independent execution.

Jaewoo-Kim-Rice committed:
handling CUDA error

imcbrearty commented:
I like this fix, and it makes sense, but could you confirm that when this is breaking for you it's true that t_index == l_dt (or 127360 in your example), rather than t_index > l_dt by a large amount? The reason is, this type of bug could also indicate something larger than just the slight numerical rounding/precision differences of the GPUs, and more something related to problems in the travel times. This fix will hide the issue if it's actually one caused by the travel times. I doubt that's the case, as that type of problem hasn't occurred in a while, but I would just like to confirm.

Jaewoo-Kim-Rice commented:
I agree that this edge case need to be invested deeper. Here I printed out all the debug printings... Any thoughts?
```
dt=0.069999993, dt0=-0.700000, dtN=2.730000
tpick(min,max)=-1.998710,3.968188 
n_sta=215, n_temp=500, k_infer=10, l_dt=50, n_arvs=204
t_index, l_dt: tensor([-12,  18,  23,  29,  44,  61,   9,   0,  56,  -1,  59,   7,  62,  -2,
         55, -10,   7,  -2,  60,  -5,  12,  25,  49,  15,  31,  -7,  53, -16,
         55,  55,  23,  31,   3,  27,  20,  -6,  34,  27,  27,  23,  25,  27,
         23,  23,  27,  22,  22,  22,  26, -10,  26,  23,  26, -15,  22,  26,
         33,  27,  22,  26,  33,  23,  26,  26,  25,  11,  22,  22,  26,  66,
         22,  25, -15,  23,  26, -16,  22,  25,  22,  25,  40,  55,  22,  26,
         25,   4,  22,  22,  25,  26,  22,  25,  34,  22,  22,  25,  31,  62,
         22,  25,  22,  25,  43,  22,  25,  22,  25,  -8,  20,  22,  25,   4,
         21,  22,  22,  25,  54,  22,  25,  49,  22,  25,  24,  55,  22,  25,
         22,  25,  21,  23,  24,  21,  24,  40,  22,  25,  28,  21,  25,  21,
         24,  10,  21,  25, -19,  21,  25,  21,  21,  24, -14,  21,  24,  21,
         24,  21,  24,  21,  24,   0,  21,  24,  21,  24,  21,  24,  21,  24,
         21,  24,  21,  24,  21,  24,  36,  24,  21,  23,  40,  21,  24,  24,
         21,  24,  21,  22,  23,  21,  21,  24,  21,  23,  38,  23,  21,  21,
         23,  21,  23,  21,  23,  60,  21,  51]) 50
=== DEBUG INFO ===
A_edges shape: torch.Size([107500]), dtype: torch.int64
t_index shape: torch.Size([2040]), dtype: torch.int64
t_index min: 5380, max: 107519
Expected A_edges size (dim 0): 107500
tpick sample: [-1.5239991   0.56313473  0.969672    1.3419753   2.4030936 ]
dt_partition: tensor([-7.0000e-01, -6.3000e-01, -5.6000e-01, -4.9000e-01, -4.2000e-01,
        -3.5000e-01, -2.8000e-01, -2.1000e-01, -1.4000e-01, -7.0000e-02,
        -4.4409e-16,  7.0000e-02,  1.4000e-01,  2.1000e-01,  2.8000e-01,
         3.5000e-01,  4.2000e-01,  4.9000e-01,  5.6000e-01,  6.3000e-01,
         7.0000e-01,  7.7000e-01,  8.4000e-01,  9.1000e-01,  9.8000e-01,
         1.0500e+00,  1.1200e+00,  1.1900e+00,  1.2600e+00,  1.3300e+00,
         1.4000e+00,  1.4700e+00,  1.5400e+00,  1.6100e+00,  1.6800e+00,
         1.7500e+00,  1.8200e+00,  1.8900e+00,  1.9600e+00,  2.0300e+00,
         2.1000e+00,  2.1700e+00,  2.2400e+00,  2.3100e+00,  2.3800e+00,
         2.4500e+00,  2.5200e+00,  2.5900e+00,  2.6600e+00,  2.7300e+00])
===================

  File "/home/jk103/GENIE/Code/module.py", line 679, in forward
    sliced_edges = torch.cat((A_edges[t_index].reshape(1,-1), src_index), dim = 0).contiguous().long()
IndexError: index 107510 is out of bounds for dimension 0 with size 107500
```

Jaewoo-Kim-Rice commented:
Potential problem could be coming from `dt_partition` discrepancy
```
dt_partition in module.py:  ['-0.699999988', '-0.629999995' ...]
dt_partition in train_GENIE_model.py:  ['-0.700000000', '-0.630000000' ...]
```

Jaewoo-Kim-Rice commented:
> Potential problem could be coming from `dt_partition` discrepancy
> 
> ```
> dt_partition in module.py:  ['-0.699999988', '-0.629999995' ...]
> dt_partition in train_GENIE_model.py:  ['-0.700000000', '-0.630000000' ...]
> ```

Fixed this issue but still getting same error.

Jaewoo-Kim-Rice commented:
I found that this error coming from discrepancy between `kernel_sig_t` and `src_t_kernel` I was using train_config that was 
`
kernel_sig_t: 0.35;
src_t_kernel: 1.0
`
And this was causing `tpick` values to be not in the range of `dt_partition`. 

When I change those two values identical, it works on current envrionment without clamping.

But not sure why this error was not raised on previous GPU environment.

Jaewoo-Kim-Rice committed:
Fix dt_partition precision loss from GPU conversion

Fix numerical precision issue where dt_partition values were being
altered due to GPU tensor conversion and back to CPU. This caused
mismatches between train_GENIE_model.py and module.py calculations.

Changes:
- module.py: Extract dt directly from numpy array to avoid GPU precision loss
  (dt = float(dt_partition[1] - dt_partition[0]))
- train_GENIE_model.py: Keep dt_partition as numpy array instead of
  converting to torch.Tensor

This ensures dt_partition maintains exact precision (e.g., -0.700000000
instead of -0.699999988) across different GPU/CUDA environments.

imcbrearty commented:
Great, I'm glad it works now. This is one of those updates I was mentioning to the hyper-parameters - more of those parameters are now automatically set and proportional to src_t_kernel, but maybe in your old versions of the codes that hadn't been updated. That should make a difference on the model behavior as well once you update those and synchronize them. Thanks for tracking it down. It is odd the other GPU didn't cause an issue - it's unclear why. But at least it's now working without a bug. In that case I probably won't merge this clamp update, since this type of error usually indicates something else is wrong (like those parameters). Though in general they also shouldn't have to be equal by the way - so you might still have something different in your versions
